[utils] add file extension provider to enclose extension handling (AS + binary addons)

xhaggi opened the pull request:
<!--- Provide a general summary of your change in the Title above -->
This PR adds a new `FileExtensionProvider` to enclose extension providing from `AdvancedSettings` and binary addons like `vfs`, `encoder`.

@notspiff mind taking a look too.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
If vfs.rar is not installed or disabled Kodi should not support `.rar` files. They should not to be scanned by our info scanners, because it results in various issues if we only add `file.rar` to the library instead of `rar://path-to-rar/file.mkv` (dupe entries, not able to playback etc.)

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
Runtime tested on MacOS.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

notspiff commented:
no time right now, but .001 is also used on vcd's so i'd be a bit careful with removing that one (though vcd isn't exactly state of the art these days).

xhaggi commented:
Okay thanks, will re-add .001 if that is used by vcd.

AlwinEsch commented on xbmc/Application.cpp:
```diff
@@ -1211,6 +1211,9 @@ bool CApplication::Initialize()
   CServiceBroker::GetRepositoryUpdater().Start();
   CServiceBroker::GetServiceAddons().Start();
 
+  CAddonMgr::GetInstance().Events().Subscribe(this, &CApplication::OnAddonEvent);
```
A `CServiceBroker::GetAddonMgr().Events().Unsubscribe(this);` is missing.

AlwinEsch replied:
Can `Subscribe` be done from CAdvancedSettings itself?

xhaggi replied:
As is init before app is init, tried before, so no.

xhaggi replied:
We could expose a method that do subscribe/unsubscribe but need to call it in `CApplication` let me know what you prefer.

tamland replied:
I prefer this to be removed from both CApplication and CAdvancedSettings, and rather made a separate 'service' that holds the extension info. It can then read extensions from both AS and addons. Others that needs the combined list reads from it. 

URIUtils::RemoveExtension should be moved here too to get rid of the dependency on globals.

xhaggi replied:
This will end up in a massive refactor because we use AS's `GetXXXExtensions` everywhere.

tamland replied:
It's just a matter of replacing `g_advancedSettings.GetXXXExtensions` with a call to another singleton. If that require too much refactoring, just have `GetXXXExtensions` forward the calls as a temp solution; that's not an issue.

The important part is to start replacing these ugly mutable globals with things that have a clear lifetime.

xhaggi replied:
Sure. I'm a bit lazy, because my spare time is limited ;)

AlwinEsch commented on xbmc/settings/AdvancedSettings.cpp:
```diff
@@ -1427,29 +1431,62 @@ void CAdvancedSettings::setExtraLogLevel(const std::vector<CVariant> &components
+    CServiceBroker::GetBinaryAddonManager().GetAddonInfos(addonInfos, true, type);
+    for (auto addonInfo : addonInfos)
+    {
+      std::string ext = addonInfo->Type(type)->GetValue(ADDON_VFS ? "@extensions" : "@extension").asString();
```
there is by `ADDON_VFS ? ...` a `... == type` missing

AlwinEsch replied:
Maybe possible to change the `@extension` to `@extensions` on the types?

xhaggi replied:
Ahh thx will address it asap

xhaggi replied:
>Maybe possible to change the @extension to @extensions on the types?

Not sure why this differs for various types. But would be good to unify it.

notspiff replied:
one of these things that happens by accident. sync it up!

notspiff commented:
looks good to me, apart from what has been discussed.

xhaggi committed:
[utils] add file extension provider to enclose extension handling (AS + binary addons)

xhaggi commented:
@tamland here we go, started from scratch and introduced a new `ExtensionManager` as you prefer, which enclode the magic for both AS + binary addon extensions.

xhaggi commented:
BTW ignore the first commits, only the last is relevant, because this PR depends on #12843.

AlwinEsch commented on xbmc/utils/ExtensionManager.cpp:
```diff
@@ -0,0 +1,142 @@
+  std::vector<std::string> extensions;
+  BinaryAddonBaseList addonInfos;
+  CServiceBroker::GetBinaryAddonManager().GetAddonInfos(addonInfos, true, type);
+  for (auto addonInfo : addonInfos)
```
You can use a `for (const auto& addonInfo : addonInfos)` to prevent a class copy

xhaggi replied:
sure

xhaggi replied:
done

AlwinEsch commented:
Only the small thing the others looks very good for me and approved from my side.

tamland commented on xbmc/utils/ExtensionManager.cpp:
```diff
@@ -0,0 +1,142 @@
+  if (typeid(event) == typeid(AddonEvents::Disabled) ||
+      typeid(event) == typeid(AddonEvents::Unload) ||
+      typeid(event) == typeid(AddonEvents::Enabled) ||
+      typeid(event) == typeid(AddonEvents::Load))
```
As mentioned in the other pr: don't use Load/Unload. This is for something else entirely

xhaggi replied:
ok removed load/unload

garbear commented:
LGPL?

xhaggi commented:
@garbear what do you mean?

notspiff commented:
License header in new file..

xhaggi commented:
Ahh c/p from another file. Will change it.

MartijnKaijser commented:
Might need to "fix" that other file?

xhaggi commented:
Puuh not sure anymore from which one I've copied the header ;)

Rechi commented:
`git grep 'GNU Lesser General' -- xbmc` lists 19 files.

xhaggi commented:
Suee sure but would go out of scope.of this PR

ksooo commented:
@xhaggi I would use "FilenameExtension" for naming here, not just "Extension". The latter is imo too generic. There can be other things that have extensions. Just my 2 ct, though.

xhaggi commented:
@ksoo sure. Thought about a good class name, but end up with this one :) .. Not sure if we really need the `Manager` suffix .. I would go with `FilenameExtensionService`.

garbear commented:
Or `FileExtensionService` for brevity

xhaggi commented:
@garbear sure .. what you think about `FileExtensionProvider`

garbear commented:
If the choice is between `manager`, `service` and `provider`, I think I like manager the best.

Thanks for this, it looks like something I could relocate my game extension stuff to. BTW, for game extensions I actually use extensions from remote non-downloaded add-ons. This allows for "just-in-time installation", where Kodi can install the appropriate game client without any effort on the user's part.

Also, HAHA I think the conflict is my fault ;)

xhaggi commented:
Updated. Addressed the header and @garbear I decided to use `FileExtensionProvider` because you can't manage anything. It only provides various filename extensions.

xhaggi commented:
@alanwww1 @tamland mind taking a look again.

garbear commented on xbmc/ServiceManager.cpp:
```diff
@@ -188,6 +191,7 @@ void CServiceManager::DeinitStageTwo()
   m_binaryAddonManager.reset();
   m_addonMgr.reset();
   m_Platform.reset();
+  m_fileExtensionProvider.reset();
```
This needs to go at the beginning of the function. Deinit in the opposite order as init to fix cross dependency issues

xhaggi replied:
Thx, will change the order

xhaggi commented:
any objections?

jenkins build this please

xhaggi commented:
jenkins build this please again

xhaggi commented:
jenkins build this please
